app/src/main/java/com/example/meetup/Fragments/SearchFragment.java

```diff
@@ -0,0 +1,391 @@
+                    String textCaption = post.getCaption();
+                    analyzeTextUsingCloudNLPApi(textCaption);
+
+                }
+                searchNameAdapter.notifyDataSetChanged();
+                searchCategoriesAdapter.notifyDataSetChanged();
+                searchRolesAdapter.notifyDataSetChanged();
+            }
+        });
+
+        setAdaptersToViews();
+    }
+
+    public void setAdaptersToViews() {
+        searchName.setAdapter(searchNameAdapter);
+        searchName.setThreshold(1);
+        searchCategory.setAdapter(searchCategoriesAdapter);
+        searchCategory.setThreshold(1);
+        searchRole.setAdapter(searchRolesAdapter);
+        searchRole.setThreshold(1);
+        searchKeyWord.setAdapter(searchKeywordsAdapter);
+        searchKeyWord.setThreshold(1);
+    }
+
+    private void queryPosts(String searchName, String searchCategory, String searchKeyWord, String searchRole) {
+        ParseQuery<Post> query = ParseQuery.getQuery(Post.class);
+        query.include(Post.KEY_USER);
+        // search for key work both in short caption and description
+        // if the key word appears in any of them, add to query result
+        if (searchKeyWord != null && searchKeyWord.length() > 0) {
+            //query.whereContains(Post.KEY_DESCRIPTION, searchKeyWord);
+            ParseQuery<Post> queryCaption = ParseQuery.getQuery(Post.class);
+            queryCaption.whereContains(Post.KEY_CAPTION, searchKeyWord);
+            ParseQuery<Post> queryDescription = ParseQuery.getQuery(Post.class);
+            queryDescription.whereContains(Post.KEY_DESCRIPTION, searchKeyWord);
+
+            List<ParseQuery<Post>> queries = new ArrayList<ParseQuery<Post>>();
+            queries.add(queryCaption);
+            queries.add(queryDescription);
+
+             query = ParseQuery.or(queries);
+             query.include(Post.KEY_USER);
+        }
+        if (searchName != null && searchName.length() > 0)
+            query.whereEqualTo(Post.KEY_STARTUP_NAME, searchName);
+        if (searchCategory != null && searchCategory.length() > 0)
+            query.whereEqualTo(Post.KEY_CATEGORY, searchCategory);
+        if (searchRole != null && searchRole.length() > 0)
+            query.whereContains(Post.KEY_ROLES, searchRole);
+
+        query.findInBackground(new FindCallback<Post>() {
+            @Override
+            public void done(List<Post> posts, ParseException e) {
+                if (e != null) {
+                    Log.e(TAG, "issue with getting posts");
+                }
+                for (Post post: posts) {
+                    Log.i(TAG, "Post " + post.getDescription() + " username " + post.getUser().getUsername());
+                }
+
+                // save received posts to list and notify adapter of new data
+                allPosts.addAll(posts);
+                adapter.notifyDataSetChanged();
+            }
+        });
+    }
+
+    private void getColumnArray(Function callback) {
```

Comment:
Suggest: there seems duplicate part between getColumnArray and queryPosts. I think we can just keep queryPosts with callback.

Reply:
the query is different, in query posts i only want select posts while in getColumnArray i want all the entries. I'm not sure how to get rid of the duplicate code